jerry-core/ecma/operations/ecma-promise-object.c

```diff
@@ -0,0 +1,719 @@
+  /* 3. */
+  ecma_value_t already_resolved = ecma_op_object_get (function_p, str_1);
+
+  ecma_value_t ret;
+  /* 4. */
+  if (ecma_is_already_resolved_value_true (already_resolved))
+  {
+    ret = ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+    goto end_of_resolve_function;
+  }
+  /* 5. */
+  ecma_set_already_resolved_value (already_resolved, true);
+
+  /* 6. */
+  if (argv[0] == promise)
+  {
+    ecma_object_t *error_p = ecma_new_standard_error (ECMA_ERROR_TYPE);
+    ret = reject_promise (promise, ecma_make_object_value (error_p));
+    ecma_deref_object (error_p);
+    goto end_of_resolve_function;
+  }
+  /* 7. */
+  if (!ecma_is_value_object (argv[0]))
+  {
+    ret = fulfill_promise (promise, argv[0]);
+    goto end_of_resolve_function;
+  }
+  /* 8. */
+  ecma_string_t *str_then = ecma_new_ecma_string_from_magic_string_id (LIT_MAGIC_STRING_THEN);
+  ecma_value_t then = ecma_op_object_get (ecma_get_object_from_value (argv[0]), str_then);
+
+  if (ECMA_IS_VALUE_ERROR (then))
+  {
+    /* 9. */
+    ret = reject_promise (promise, then);
+  }
+  else if (!ecma_op_is_callable (then))
+  {
+    /* 11 .*/
+    ret = fulfill_promise (promise, argv[0]);
+  }
+  else
+  {
+    /* 12 */
+    ecma_enqueue_promise_resolve_thenable_job (promise, argv[0], then);
+    ret = ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+  }
+
+  ecma_deref_ecma_string (str_then);
+  ecma_free_value (then);
+
+end_of_resolve_function:
+  ecma_deref_ecma_string (str_0);
+  ecma_deref_ecma_string (str_1);
+  ecma_free_value (promise);
+  ecma_free_value (already_resolved);
+  return ret;
+} /* ecma_promise_resolve_handler */
+
+/**
+ * Native handler for CapabilityiesExecutor Function.
+ *
+ * See also: ES2015 25.4.1.5.1
+ *
+ * @return ecma value
+ */
+static ecma_value_t
+ecma_promise_executor_handler (const ecma_value_t function,
+                               const ecma_value_t this,
+                               const ecma_value_t argv[],
+                               const ecma_length_t argc)
+{
+  JERRY_UNUSED (this);
+  JERRY_UNUSED (argc);
+
+  ecma_string_t *str_0 = ecma_new_ecma_string_from_uint32 (0);
+  ecma_string_t *str_1 = ecma_new_ecma_string_from_uint32 (1);
+  ecma_string_t *str_2 = ecma_new_ecma_string_from_uint32 (2);
+  ecma_value_t ret = ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+  ecma_value_t reject = ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+  /* 2. String '0' indicates [[Capability]] of the function object. */
+  ecma_value_t capability = ecma_op_object_get (ecma_get_object_from_value (function), str_0);
+  /* 3. String '1' indicates [[Resolve]] of the capability. */
+  ecma_value_t resolve = ecma_op_object_get (ecma_get_object_from_value (capability), str_1);
+
+  if (resolve != ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED))
+  {
+    ret = ecma_raise_type_error (ECMA_ERR_MSG ("'resolve' function should be undefined."));
+    goto end_of_executor_function;
```

Comment:
yes, I can remove this `goto`.
What about the `goto` in `ecma_promise_resolve_handler `? I think that one is needed, otherwise the code will be quite 'dirty'